### PR TITLE
docs(headless): add React samples for headless facet

### PR DESCRIPTION
**Pull request by btaillon**
https://coveord.atlassian.net/browse/KIT-392

**Warning**: This PR is not intended to be merged. It's here for discussion.

Prior to the recent changes in this PR, this error appeared:
image

Turns out that the cause was that `<React.StrictMode></React.StrictMode>` wrapped the whole app, which causes every class component's constructor along with some other state functions to be executed twice. To quote [their unit tests](https://github.com/facebook/react/blob/702fad4b1b48ac8f626ed3f35e8f86f5ea728084/packages/react/src/__tests__/ReactStrictMode-test.js#L106):
```js
expect(log).toEqual([
    'constructor',
    'constructor',
    'getDerivedStateFromProps',
    'getDerivedStateFromProps',
    'render',
    'render',
    'componentDidMount',
]);
```

While removing `StrictMode` or running in production removes the error, neither approach seemed advisable. Additionally, this unveiled the possible issue that a component whose constructor may be executed again after removal may cause issues.

This PR attempts to solve that (for testing & discussion purpose) by introducing a way to remove/clean facets.

Here are some questions I'd like to discuss:
* Should we introduce an action to clean facets?
  * What should it be named?
  * Should it automatically be called by the controller when unsubscribing?
  * Should more components feature a way to clean them?
  * Should it be introduced in this PR or a different one?

This PR doesn't include `facet-search`.

### CODE REVIEW

File: packages/samples/headless-react/src/components/facet/facet.class.tsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  public state: FacetState;
+  private unsubscribe: Unsubscribe = () => {};
+
+  constructor(props: FacetProps) {
+    super(props);
+    if (process.env.NODE_ENV === 'development') {
+      // When `React.StrictMode` is active in development, the constructor is executed twice.
+      // This ensures the facet isn't already defined from a previous execution.
+      engine.dispatch(FacetActions.removeFacet(this.props.facetId));
+    }
```
Comment (btaillon) [inline packages/samples/headless-react/src/components/facet/facet.class.tsx:27]: Unfortunately, it looks like after the first execution of the constructor, every private attribute is set to `undefined`.

**Comment by ThibodeauJF** [comment]
That's just a warning that is logged when a field has already been used as an ID, right? Quick solution would be to put an ID on the facet, like the message says, or to simply delete the warning, since we "fix it" by appending an id to the field anyways.

The facet won't actually register itself twice, looking at facet-set-slice, it will simply return. That was a decision we made to prevent errors when registering facets multiple times, which can happen often when using the facet manager and disconnecting/reconnecting facets to the DOM (in atomic for example).

Furthermore like you mentioned it's for react.strictmode in dev, not a huge deal, could even had a nota bene in the warning...

Removing facets could be useful but I don't think it's the issue here

**Comment by samisayegh** [comment]
#### Keeping state in store: feature or bug?

An issue with removing facets from the store on unsubscribe comes from the Atomic Facet Manager.

Reordering the facet nodes triggers the unmount lifecycle, which unsubscribes. However, if the facet was also removed from the store, then on remount, the facet would be empty.

With Atomic Facets, `subscribe` and `unsubscribe` are called many times, but `buildFacet` is called once, so we do not see the warning. But there are use-cases for calling `buildFacet` multiple times with the same field. With SSR, the build function would be called once on the server, and once on the client when the id is already in the store.

Calling a build function with the same identifier is a feature; it allows one to re-control and re-sync with a piece of state. If headless automatically cleared the piece of state on unsubscribe, headless would not be able to support certain use-cases.

#### Memory-leak risks

There are certain components in headless that can have multiple instances such as search-boxes and facets. If someone has a use-case that requires rotating through hundreds and thousands of search-boxes/facets, then headless would eventually use all the available memory. By rotating, I mean registering and discarding instances, rather than reusing an instance.

I would say > 99% of people do not have such a use-case, so there would be no need to manage unregistering state. If we encounter someone in the 1%, then we'd need to understand what is going on with their implementation. They might need e.g. an action that purges all facets, rather than a single facet by id. I think it is too early to add actions to remove instances.

#### Improving the warning

Ideally this warning would not display. The problem is when `buildFacet` is called with the same field, we do not know if it is to a) resync with the first instance, or b) to create a second instance. The warning is meant to alert the user about situation b, but here we have situation a.

The simplest solution to resolve the warning would be to pass an id so headless knows it is dealing with situation a. In the store, there would be a single registered facet, versus two.

**Comment by btaillon** [comment]
Perfect, then I'll add a `facetId` and revert the addition of this unnecessary action.

**Pull request closed without merging**